src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs

```diff
@@ -2009,6 +2009,102 @@ public static void Sort<T>(this Span<T> span, Comparison<T> comparison)
+        {
+            // Shrink one of the spans if necessary to ensure they're both the same length. We can then iterate until
+            // the Length of one of them and at least have bounds checks removed from that one.
+            if (other.Length > span.Length)
+            {
+                other = other.Slice(0, span.Length);
+            }
+            else if (span.Length > other.Length)
+            {
+                span = span.Slice(0, other.Length);
+            }
+            Debug.Assert(span.Length == other.Length);
+
+            // Find the first element pairwise that is not equal, and return its index as the length
+            // of the sequence before it that matches.
+            for (int i = 0; i < span.Length; i++)
+            {
+                if (!EqualityComparer<T>.Default.Equals(span[i], other[i]))
+                {
+                    return i;
+                }
+            }
+
+            return span.Length;
+        }
+
+        /// <summary>Determines the length of any common prefix shared between <paramref name="span"/> and <paramref name="other"/>.</summary>
+        /// <typeparam name="T">The type of the elements in the sequences.</typeparam>
+        /// <param name="span">The first sequence to compare.</param>
+        /// <param name="other">The second sequence to compare.</param>
+        /// <param name="comparer">The <see cref="IEqualityComparer{T}"/> implementation to use when comparing elements, or null to use the default <see cref="IEqualityComparer{T}"/> for the type of an element.</param>
+        /// <returns>The length of the common prefix shared by the two spans.  If there's no shared prefix, 0 is returned.</returns>
+        public static int CommonPrefixLength<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other, IEqualityComparer<T>? comparer)
+        {
+            // Shrink one of the spans if necessary to ensure they're both the same length. We can then iterate until
+            // the Length of one of them and at least have bounds checks removed from that one.
+            if (other.Length > span.Length)
+            {
+                other = other.Slice(0, span.Length);
+            }
+            else if (span.Length > other.Length)
+            {
+                span = span.Slice(0, other.Length);
+            }
+            Debug.Assert(span.Length == other.Length);
+
+            // Find the first element pairwise that is not equal, and return its index as the length
+            // of the sequence before it that matches.
+            if (typeof(T).IsValueType && (comparer is null || comparer == EqualityComparer<T>.Default)) // special-cased to enable devirtualization
+            {
+                for (int i = 0; i < span.Length; i++)
+                {
+                    if (!EqualityComparer<T>.Default.Equals(span[i], other[i]))
+                    {
+                        return i;
+                    }
+                }
```

Comment:
is there any reason for why we should not be calling `CommonPrefixLength<T>` here?

```suggestion
                return CommonPrefixLength<T>(span, other);
```

Reply:
Not wanting to pay the overhead again for the upfront slicing and additional method invocation.  I debated moving the check earlier to be the first thing in the method and delegating in that case, but it would still incur the additional method call.  But, it's probably not a huge deal, and passing in a comparer is not the common case.